src/control/jobs/XournalScheduler.cpp

```diff
@@ -12,51 +12,60 @@ XournalScheduler::XournalScheduler()
 XournalScheduler::~XournalScheduler() = default;
 
 void XournalScheduler::removeSidebar(SidebarPreviewBaseEntry* preview) {
-    // Dont await for tasks, as it completly blocks whole program.
-    // Not sure why.
-    removeSource(preview, JOB_TYPE_PREVIEW, JOB_PRIORITY_HIGH, false);
+    removeSource(preview, JOB_TYPE_PREVIEW, JOB_PRIORITY_HIGH, true);
 }
 
 void XournalScheduler::removePage(XojPageView* view) { removeSource(view, JOB_TYPE_RENDER, JOB_PRIORITY_URGENT); }
 
 void XournalScheduler::removeAllJobs() {
-    g_mutex_lock(&this->jobQueueMutex);
+    std::lock_guard<std::mutex> lock(this->jobQueueMutex);
 
     for (int priority = JOB_PRIORITY_URGENT; priority < JOB_N_PRIORITIES; priority++) {
-        int length = g_queue_get_length(this->jobQueue[priority]);
-        for (int i = 0; i < length; i++) {
-            Job* job = static_cast<Job*>(g_queue_peek_nth(this->jobQueue[priority], i));
+        std::deque<Job*>& queue = *this->jobQueue[priority];
+        auto it = queue.begin();
 
+        while (it != queue.end()) {
+            Job* job = *it;
+
+            // Only remove PREVIEW and RENDER jobs; we aren't
+            // responsible for other types of jobs.
             JobType type = job->getType();
             if (type == JOB_TYPE_PREVIEW || type == JOB_TYPE_RENDER) {
                 job->deleteJob();
-                g_queue_remove(this->jobQueue[priority], job);
+
+                it = queue.erase(it);
+
                 job->unref();
+                job = nullptr;
+            } else {
+                ++it;
```

Comment:
Use `std::erase` (cpp20) or `std::remove_if` in combination with `container.erase`

Reply:
To me,
```c++
        queue.erase(std::remove_if(queue.begin(), queue.end(),
            [&](Job* job) {
                if (job->getType() == type && job->getSource() == source) {
                    job->deleteJob();
                    job->unref();

                    return true;
                }   

                return false;
            }), queue.end());
```

makes less sense than

```c++
        auto it = this->jobQueue[priority]->begin();

        while (it != this->jobQueue[priority]->end()) {
            Job* job = *it;

            if (job->getType() == type && job->getSource() == source) {
                it = this->jobQueue[priority]->erase(it);

                job->deleteJob();
                job->unref();
                job = nullptr;
            } else {
                ++it;
            }   
        } 
```

because we're calling `job->unref()` in the predicate given to `std::remove_if`. I'm willing to change it to the `queue.erase(std::remove_if(...))` version, if you disagree, though.

Reply:
We need a refactor of the job system. It should use `std::shared_ptr`s. The actions executed here have to be done for each destruction of Jobs, so they should be moved into the destructor.
Basically `Job::deleteJob()` must be removed. And `Job::unref` has to be handled by a `shared_ptr`.

Reply:
Regarding the while over `remove_if` question. This is indeed ideological. 
In my opinion, a loop whether `while` or `for` does not tell anything about what happens there.

`remove_if` exactly tells the reader, that you want to erase objects from the container.
Also, removing elements from the list while iterating is first prone to bugs, second it can be slow for some containers. (Not in case of deque's).

> I would very much like the Job* job parameter to be a const Job* job in the lambda
Yes, I agree, that the parameter should be `const`, but that's not possible, because the `Job`s are designed wrongly.

> I don't think remove_if guarantees that the predicate will be called exactly once for each job

Even if not: 2*O(1) is still O(1) and 2*O(n) is still O(n) (2*O(n²) = O(n²))...
So the disadvantage would be minimal, but the `erase(remove_if` is designed to make deleting many objects from the list as optimal as possible. It works by moving all the objects to be removed to the end of the list. Then the complete end is removed at once.

Reply:
Ok just read the [documentation](https://en.cppreference.com/w/cpp/algorithm/remove): You are right, 
```

p | - | unary predicate which returns ​true  if the element should be removed.  The expression p(v) must be convertible to bool for every argument v of type (possibly const) VT, where VT is the value type of ForwardIt, regardless of value category, and must not modify v. Thus, a parameter type of VT&is not allowed, nor is VT unless for VT a move is equivalent to a copy (since C++11). ​
-- | -- | --
```